src/mpi/datatype/dataloop/dataloop_create.c

```diff
@@ -34,6 +34,13 @@ void MPIR_Dataloop_create(MPI_Datatype type, MPIR_Dataloop ** dlp_p, MPI_Aint *
     MPI_Aint *disps;
     MPI_Aint *blklen;
 
+    if (type == MPI_FLOAT_INT || type == MPI_DOUBLE_INT ||
+        type == MPI_LONG_INT || type == MPI_SHORT_INT ||
+        type == MPI_LONG_DOUBLE_INT || type == MPI_2INT) {
+        MPIR_Dataloop_create_pairtype(type, dlp_p, dlsz_p);
+        return;
+    }
+
```

Comment:
The intent was that `MPIR_Dataloop_create` be the public interface for the dataloops code.  `MPIR_Dataloop_create_vector`, `MPIR_Dataloop_create_struct`, `MPIR_Dataloop_create_pairtype`, etc., are supposed to be internal functions that `MPIR_Dataloop_create` would call based on the type of datatype used.  We were using that model correctly for all datatypes, except pairtypes.  This function makes that consistent.

With respect to ARMCI-MPI or OSHMPI, if you are creating derived datatypes frequently in the performance critical path, you have bigger problems than that one extra function-call overhead.  :-)  Compared to the total cost of that function, the extra function call would be in the noise.

Reply:
Unlike `vector|struct...`, each pairtype's data loop structure never changes. It somehow makes sense to me that we handle pairtype's data loop separately. I agree that the additional overhead is small, but it reads like we add unnecessary overhead to every creation call just for handling static data loop structures.

For the public interface aspect, I do not have strong preference with either option.

Reply:
@minsii I agree with you in general that pairtypes have some potential for better optimization, but I want to make two points:

1. We don't special case them today.  The code path is the same as other derived datatypes, which is very high overhead.

2. Pair types are not easy to optimize because the MPI standard requires us to not touch the padding bytes, which makes it noncontiguous.  Nathan had a proposal at the MPI Forum to remove the padding byte protection, at which point we might be able to do a fastpath for pairtypes.  But even then the fastpath would essentially make pairtypes as "builtin" types, rather than a derived datatype.  So this change wouldn't matter.

Finally, I want to remind you that here you are arguing for one extra function call and one branch.  That's it.  There's no additional overhead that's being added.  I'd be shocked if you can find a system where this overhead would be noticeable compared with the cost of the `MPIR_Dataloop_create_pairtype` function itself.

For the public interface, I'm not sure what "either option" is here.  The intent was always that `MPIR_Dataloop_create` be the public interface and everything else be a private interface.  Somehow someone patched a hole in that interface (maybe because the namespace was not cleanly done) and called the `MPIR_Dataloop_create_pairtype` function outside of the dataloop code.  There are other such examples too, which I'm slowly fixing.

Reply:
Sorry, I might have misunderstood your concern a bit here.  You are arguing for the cost of `MPIR_Dataloop_create`, not that of `MPIR_Dataloop_create_pairtype`, aren't you?  In that case, the cost is just this one branch that'll never be taken in your example:

```
    if (type == MPI_FLOAT_INT || type == MPI_DOUBLE_INT ||
         type == MPI_LONG_INT || type == MPI_SHORT_INT ||
         type == MPI_LONG_DOUBLE_INT || type == MPI_2INT) {
         MPIR_Dataloop_create_pairtype(type, dlp_p, dlsz_p);
         return;
    }
```

Is that really a concern?  This is during the actual datatype creation, not during communication.  I understand that in ARMCI-MPI, the datatype creation is done in the performance critical path, but the cost of this branch would be negligible even for very small messages.  I find it to be a very weak reason to poke holes in the public interface as an optimization path.

Reply:
This condition check is always performed at every `MPIR_Dataloop_create` call. It may take at least 6 compare instructions and 1 branch. Anyway, I agree that the overhead should be small comparing to the other part of this call.

You can add my signoff in this commit.

Reply:
No, that's not true.  Any decent compiler worth its salt will convert that to a labeled jump statement (similar to what you'd get with a switch statement).  So it doesn't depend on the number of checks.